internal/image/history.go

```diff
@@ -89,25 +93,91 @@ func (h ImageHistory) load(ctx context.Context, name reference.Named, digest dig
 		h.byName[key] = bucket
 	}
 
+	bucket.entries = append(bucket.entries, entry)
+	if entry.After(bucket.mostRecent.CheckpointID) {
+		bucket.mostRecent = entry
+	}
+}
+
+func (h ImageHistory) load(
+	ctx context.Context,
+	name reference.Named,
+	digest digest.Digest,
+	checkpoint CheckpointID,
+	service model.Service,
+) (refKey, historyEntry, error) {
+	h.mu.Lock()
+	defer h.mu.Unlock()
+
+	key := makeRefKey(name)
+	bucket, ok := h.byName[key]
+	if !ok {
+		bucket = &NamedImageHistory{name: name}
+		h.byName[key] = bucket
+	}
+
+	hi, err := hashService(service)
+	if err != nil {
+		return "", historyEntry{}, err
+	}
+
 	entry := historyEntry{
-		Digest:       digest,
-		CheckpointID: checkpoint,
+		Digest:        digest,
+		CheckpointID:  checkpoint,
+		HashedService: hi,
 	}
 	bucket.entries = append(bucket.entries, entry)
 	if entry.After(bucket.mostRecent.CheckpointID) {
 		bucket.mostRecent = entry
 	}
 
-	return key, entry
+	return key, entry, nil
+}
+
+type hash struct {
+	BaseDockerfile build.Dockerfile
+	Mounts         []model.Mount
+	Steps          []model.Cmd
+	Entrypoint     model.Cmd
 }
 
-func (h ImageHistory) AddAndPersist(ctx context.Context, name reference.Named, digest digest.Digest, checkpoint CheckpointID) error {
-	key, entry := h.load(ctx, name, digest, checkpoint)
+type HashedService = uint64
+
+func hashService(service model.Service) (HashedService, error) {
+	hi := hash{
+		BaseDockerfile: build.Dockerfile(service.DockerfileText),
```

Comment:
this is a dangerous approach. if someone adds new fields to Service, they won't know to update them here.

can you either:
1) hash the existing Service structure, but void out the fields you don't care about?
2) create some sort of hash method on Service itself?

Reply:
ooohhh hash method on the service++

Reply:
yeah that's a great point @nicks, I like that idea.